File: Mobile/NFCVerifier_iOS/Sources/Services/AppFileDecoder/AppFileDecoder.swift
/*
 * Copyright 2022 Infineon Technologies AG.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import Foundation

/**
 * Base class defining the decoders for product and service information received from the tag
 */
class AppFileDecoder {
    
	/**
     * Parses the custom data field
     *  - Parameter payload: Product or service information payload
	 *  - Parameter index: Index of the custom data field
     *  - Returns the decoded custom data in key-value format
     */
    static func parseCustomField(payload: Data, index: inout Int) -> [CustomFieldItem] {
        let range = getCustomFieldRange( byte: payload, index: &index)
        let data = String(bytes: payload.subdata(in: range), encoding: .utf8)!
        let itemArr = data.split(separator: ";")
        var counter = 0
        var customFieldItems = [CustomFieldItem]()
        while(counter<itemArr.count){
            let substring = itemArr[counter].split(separator: ":")
            customFieldItems.append(CustomFieldItem(title: String(substring[0]), value: String(substring[1])))
            counter += 1
        }
        return customFieldItems
    }
	
	/**
     * Returns the range of the custom field in the payload byte array
     *  - Parameter payload: Product or service information payload
	 *  - Parameter index: Index of the custom data field
     *  - Returns the range of the custom data field
     */
    static func getCustomFieldRange(byte:Data, index: inout Int) -> Range<Data.Index>  {
        let data = byte.subdata(in: index..<(index+2))
        let customFieldLen = Utils.byteToShort(data: data)
        let range: Range<Data.Index> = index+2..<(index+2+customFieldLen)
        return range
    }
    
	/**
     * Parses the date fields
     *  - Parameter payload: Product or service information payload
	 *  - Parameter index: Index of the date field
     *  - Returns the decoded date
     */
    static func parseDateBytes(payload: Data, index: inout Int) -> Date? {
        let range :Range<Data.Index>  = index..<index+3
        let data:Data = payload.subdata(in:range)
        index += 3
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyMMdd"
        let date = dateFormatter.date(from: data.hexEncodedString())
        return date
    }    
}

File: Mobile/NFCVerifier_Android/app/src/main/java/com/infineon/nfcbpsk/services/appfiledecoder/AppFileDecoder.java
/*
 * Copyright 2022 Infineon Technologies AG.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.infineon.nfcbpsk.services.appfiledecoder;

import com.infineon.ndef.utils.Utils;

import java.io.ByteArrayInputStream;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Locale;

/**
 * Base class defining the decoders for product and service information received from the tag
 */
public class AppFileDecoder {
    final static String DATE_FORMAT = "yy MM dd";

    /**
     * Parse the date from current position to next 3 bytes
     *
     * @param bais ByteArrayInputStream pointing to the date index
     * @return return the date
     */
    protected static Date parseDateBytes(ByteArrayInputStream bais) {
        byte[] array = new byte[3];
        int index = 0;
        while (index < 3) {
            array[index] = (byte) bais.read();
            index++;
        }
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat(DATE_FORMAT, Locale.getDefault());
        try {
            String str = Utils.toHexString(array);
            return simpleDateFormat.parse(str);
        } catch (ParseException e) {
            return null;
        }
    }

    /**
     * Decodes the custom data fields from stream bytes
     *
     * @param bais ByteArrayInputStream pointing to the custom filed index
     * @return return the array list of CustomFieldItem
     */

    protected static ArrayList<CustomFieldItem> parseCustomFields(ByteArrayInputStream bais) {
        byte[] arrayLen = readBytes(bais, 2);
        int customFieldLength = Utils.getUINT16(arrayLen, 0);
        byte[] customBytes = readBytes(bais, customFieldLength);
        String customDataString = Utils.toString(customBytes, 0, customFieldLength);
        String[] listData = customDataString.split(";");
        ArrayList<CustomFieldItem> customFields = new ArrayList<>();
        for (String filed : listData) {
            String[] subData = filed.split(":");
            customFields.add(new CustomFieldItem(subData[0], subData[1]));
        }
        return customFields;
    }

    /**
     * Reads the next bytes with specified length
     *
     * @param bais   ByteArrayInputStream pointing to the current filed index
     * @param length number of bytes to read
     * @return return the byte array
     */
    protected static byte[] readBytes(ByteArrayInputStream bais, int length) {
        byte[] arrayBytes = new byte[length];
        int index = 0;
        while (index < length) {
            arrayBytes[index] = (byte) bais.read();
            index++;
        }
        return arrayBytes;
    }
}
